File: archive/2023_intern/gliacloud_intern/code_interview/ts/07.ts
function searchingChallenge(strArr: string[]): number {
    const matrix: number[][] = strArr.map(row => row.split('').map(Number)); // Convert the string matrix to a list of lists
  
    function dfs(row: number, col: number, prev: number, length: number): number {
      if (row < 0 || row >= matrix.length || col < 0 || col >= matrix[0].length) {
        return length; // Base case: out of bounds
      }
  
      const current: number = matrix[row][col];
      if (current <= prev) {
        return length; // Base case: current value is not strictly increasing
      }
  
      let maxLength: number = length;
      for (const [dx, dy] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
        const newRow: number = row + dx;
        const newCol: number = col + dy;
        maxLength = Math.max(maxLength, dfs(newRow, newCol, current, length + 1));
      }
  
      return maxLength;
    }
  
    let maxLength: number = 0;
    for (let row = 0; row < matrix.length; row++) {
      for (let col = 0; col < matrix[0].length; col++) {
        maxLength = Math.max(maxLength, dfs(row, col, -Infinity, 0));
      }
    }
  
    return maxLength;
  }
  
  console.log(searchingChallenge(prompt().split(',')));
  

File: archive/2023_intern/gliacloud_intern/code_interview/python/07.py
'''
Searching Challenge
Have the function SearchingChallenge(strArr) take the array of strings stored in strArr, which will be an NxM matrix of positive single-digit integers, and find the longest increasing path composed of distinct integers. When moving through the matrix, you can only go up, down, left, and right. For example: if strArr is ["345", "326", "221"], then this looks like the following matrix:

3 4 5
3 2 6
2 2 1

For the input above, the longest increasing path goes from: 3 -> 4 -> 5 -> 6. Your program should return the number of connections in the longest path, so therefore for this input your program should return 3. There may not necessarily always be a longest path within the matrix.
Examples
Input: ["12256", "56219", "43215"]
Output: 5
Input: ["67", "21", "45"]
Output: 3
'''

def SearchingChallenge(strArr):
    matrix = [list(map(int, row)) for row in strArr]  # Convert the string matrix to a list of lists

    def dfs(row, col, prev, length):
        if row < 0 or row >= len(matrix) or col < 0 or col >= len(matrix[0]):
            return length  # Base case: out of bounds

        current = matrix[row][col]
        if current <= prev:
            return length  # Base case: current value is not strictly increasing

        max_length = length
        for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            new_row = row + dx
            new_col = col + dy
            max_length = max(max_length, dfs(new_row, new_col, current, length + 1))

        return max_length

    max_length = 0
    for row in range(len(matrix)):
        for col in range(len(matrix[0])):
            max_length = max(max_length, dfs(row, col, float('-inf'), 0))

    return max_length

print(SearchingChallenge(input()))